src/web3/Client.ts

```diff
@@ -82,6 +94,74 @@ export class Client implements IClient {
     this.getProviders = this.getProviders.bind(this);
     this.getPrivateProviders = this.getPrivateProviders.bind(this);
     this.getPublicProviders = this.getPublicProviders.bind(this);
+    // client discovery
+    this.clientDiscoveryCallback = this.clientDiscoveryCallback.bind(this);
+    this.stopClientDiscovery = this.stopClientDiscovery.bind(this);
+    this.startClientDiscovery = this.startClientDiscovery.bind(this);
+
+    // start client discovery
+    if (withExternalProviderDiscovery) {
+      this.startClientDiscovery();
+    }
+  }
+
+  /**
+   * A callback method that triggers a ping of the Thyra's server
+   *
+   * @returns void
+   */
+  private async clientDiscoveryCallback(): Promise<void> {
+    let externalWalletProviders: IExternalWalletProvider[] | null = null;
+    try {
+      externalWalletProviders = await getExternalWalletProviders();
+    } catch (ex) {
+      console.error(`Error getting external wallet providers`);
+    }
+    if (externalWalletProviders) {
+      // add only newly discovered providers
+      // NOTE: we dont remove providers which we have been disconnected from
+      const alreadyDiscoveredProviders = new Set(
+        this.externalWalletProviders.map((p) => p.name()),
+      );
+      for (const item of externalWalletProviders) {
+        if (!alreadyDiscoveredProviders.has(item.name())) {
+          this.externalWalletProviders.push(item);
+        }
+      }
+      this.externalWalletProviders = externalWalletProviders;
+    }
+
+    // reset the interval
+    this.externalWalletProvidersDiscoveryTimeoutId = new Timeout(
+      this.externalWalletProvidersDiscoveryInterval,
+      () => this.clientDiscoveryCallback(),
+    );
+  }
+
+  /**
+   * A method to stop listening for a connection to Thyra's server
+   *
+   * @returns void
+   */
+  public stopClientDiscovery(): void {
+    if (this.externalWalletProvidersDiscoveryTimeoutId)
+      this.externalWalletProvidersDiscoveryTimeoutId.clear();
+  }
+
+  /**
+   * A method to start listening for a connection to Thyra's server.
+   *
+   * @returns void
+   */
+  public startClientDiscovery(): void {
+    const that = this;
```

Comment:
this should not be needed if you use "this" in arrow function. which is the case line 163

Reply:
Great catch, thanks!

Reply:
Fixed